hpy/devel/src/runtime/buildvalue.c

```diff
@@ -0,0 +1,145 @@
+/**
+ * Implementation of HPy_BuildValue.
+ *
+ * HPy_BuildValue creates a new value based on a format string from the values
+ * passed in an array. Returns HPy_NULL in case of an error and raises an exception.
+ *
+ * HPy_BuildValue does not always build a tuple. It builds a tuple only if its format
+ * string contains two or more format units. If the format string is empty, it returns
+ * None; if it contains exactly one format unit, it returns whatever object is described
+ * by that format unit. To force it to return a tuple of size 0 or one, parenthesize the
+ * format string.
+ *
+ * Building complex values with HPy_BuildValue is more convenient than the equivalent
+ * code that uses more granular APIs with proper error handling and cleanup. Moreover,
+ * HPy_BuildValue provides straightforward way to port existing code that uses
+ * Py_BuildValue.
+ *
+ * Supported Formatting Strings
+ * ----------------------------
+ *
+ * Numbers
+ * ~~~~~~~
+ *
+ * ``i (int) [int]``
+ *     Convert a plain C int to a Python integer object.
+ *
+ * ``f (float) [float]``
+ *     Convert a C float to a Python floating point number.
+ *
+ * Collections
+ * ~~~~~~~
+ *
+ * ``(items) (tuple) [matching-items]``
+ *     Convert a sequence of C values to a Python tuple with the same number of items.
+ *
+ * API
+ * ---
+ *
+ */
+
+#include "hpy.h"
+#include <stdarg.h>
+
+static HPy_ssize_t count_items(HPyContext *ctx, const char *fmt, char end);
+static HPy build_tuple(HPyContext *ctx, const char **fmt, va_list *values, HPy_ssize_t size, char expected_end);
+static HPy build_single(HPyContext *ctx, const char **fmt, va_list *values);
+
+HPyAPI_HELPER
+HPy HPy_BuildValue(HPyContext *ctx, const char *fmt, ...)
+{
+    va_list values;
+    va_start(values, fmt);
+    HPy_ssize_t size = count_items(ctx, fmt, '\0');
+    if (size < 0) {
+        return HPy_NULL;
+    } else if (size == 1) {
+        return build_single(ctx, &fmt, &values);
+    } else {
+        return build_tuple(ctx, &fmt, &values, size, '\0');
+    }
+}
+
+static HPy_ssize_t count_items(HPyContext *ctx, const char *fmt, char end)
```

Comment:
Since we are doing a pre-parsing of the format string here anyway:
We could add a hook to `HPy_BuildValue` that calls a context function instead of doing the parsing in C.
So, if the hook function is available, the `count_items` could then just extract the varargs into an array.
This sounds a bit awkward but in this way, we could avoid multiple upcalls while varargs unmarshalling is still done in the C extension's binary.
For CPython, the context function would just be `NULL`.
What do you think?

Reply:
I am not sure if the C specification guarantees us that something like raw memory copy of a `double` to `uint64_t` will work as expected and you'd see the right bits (in the right order) on the other side. It would probably work, but not sure if we want to rely on that.

What I think the spec guarantees is casting `double*` to `void*` and then back, but it seems that one cannot just take address of `va_arg`, so we'd have to stack allocate array for each primitive type, store all the values there and then take pointers to them to create the final `void*` array. 

Alternatively, we can just pass the individual primitive arrays to the context function, so its signature would be something like `ctx_BuildValue(const char* fmt, int[] items_i, long[] items_l, double[] items_d, HPy[] items_hpy)`. The set of supported primitive values is part of the contract in any case regardless of whether they are in the signature. This would seem like reasonable approach to me.

Having the ability to handle this sometimes complex operation efficiently in the runtime without calling back and forth between HPy and the runtime seems like useful thing for GraalPython. What about PyPy? OTOH, we could say the same about `HPyArg_Parse`, so how far do we want to go? I would be for this simple `HPy_BuildValue` for now and eventually adding the possibility to provide optimized version by the context in the future.

Reply:
Passing individual arrays sounds good to me.
I think the _hook approach_ would also be reasonable for `HPyArg_Parse`.
I'm fine if we add the hook in a later stage.